core/src/main/java/org/apache/iceberg/BaseFileScanTask.java

```diff
@@ -91,29 +93,53 @@ public String toString() {
         .toString();
   }
 
+
+  /**
+   * This iterator returns {@link FileScanTask} using guidance provided by split offsets.
+   */
   @VisibleForTesting
-  static final class OffsetsBasedSplitScanTaskIterator implements Iterator<FileScanTask> {
-    private final List<Long> splitOffsets;
+  static final class OffsetsAwareTargetSplitSizeScanTaskIterator implements Iterator<FileScanTask> {
+    private final List<Long> offsets;
+    private final List<Long> splitSizes;
     private final FileScanTask parentScanTask;
-    private int idx = 0;
-
-    OffsetsBasedSplitScanTaskIterator(List<Long> splitOffsets, FileScanTask fileScanTask) {
-      this.splitOffsets = splitOffsets;
-      this.parentScanTask = fileScanTask;
+    private final long targetSplitSize;
+    private int offsetIdx = 0;
+    private int sizeIdx = 0;
+
+    OffsetsAwareTargetSplitSizeScanTaskIterator(
+        List<Long> offsetList, FileScanTask parentScanTask, long targetSplitSize) {
+      this.offsets = ImmutableList.copyOf(offsetList);
```

Comment:
Missed this the first time: why copy the offset list? It shouldn't change.

Reply:
Just being a bit defensive. Future implementations of the DataFile interface may not create an immutable copy before passing it downstream.

Reply:
Seems fine since this should be an ImmutableList in the current read path and won't actually copy.